plugins/node/opentelemetry-koa-instrumentation/src/koa.ts

```diff
@@ -0,0 +1,155 @@
+    shimmer.unwrap(appProto, 'use');
+  }
+
+  /**
+   * Patches the Koa.use function in order to instrument each original
+   * middleware layer which is introduced
+   * @param {KoaMiddleware} middleware - the original middleware function
+   */
+  private _getKoaUsePatch(original: (middleware: KoaMiddleware) => koa) {
+    return function use(
+      this: koa,
+      middlewareFunction: KoaMiddleware,
+      ...args: Parameters<typeof original>
+    ) {
+      let patchedFunction;
+      if (middlewareFunction.router) {
+        patchedFunction = plugin._patchRouterDispatch(middlewareFunction);
+      } else {
+        patchedFunction = plugin._patchLayer(middlewareFunction, false);
+      }
+
+      args[0] = patchedFunction;
+      const res = original.apply(this, args);
+
+      return res;
+
+      // eslint-disable-next-line @typescript-eslint/no-explicit-any
+    } as any;
+  }
+
+  /**
+   * Patches the dispatch function used by @koa/router. This function
+   * goes through each routed middleware and adds instrumentation via a call
+   * to the @function _patchLayer function.
+   * @param {KoaMiddleware} dispatchLayer - the original dispatch function which dispatches
+   * routed middleware
+   */
+  private _patchRouterDispatch(dispatchLayer: KoaMiddleware) {
+    this._logger.debug('Patching @koa/router dispatch');
+
+    const router = dispatchLayer.router;
+
+    const routesStack = router?.stack ?? [];
+    for (const pathLayer of routesStack) {
+      const path = pathLayer.path;
+      const pathStack = pathLayer.stack;
+      for (let j = 0; j < pathStack.length; j++) {
+        const routedMiddleware: KoaMiddleware = pathStack[j];
+        pathStack[j] = this._patchLayer(routedMiddleware, true, path);
+      }
+    }
+    const dispatcher = (context: KoaContext, next: koa.Next) => {
+      const result = dispatchLayer(context, next);
+      return result;
+    };
+    return dispatcher;
+  }
+
+  /**
+   * Patches each individual @param middlewareLayer function in order to create the
+   * span and propagate context. It does not create spans when there is no parent span.
+   * @param {KoaMiddleware} middlewareLayer - the original middleware function.
+   * @param {boolean} isRouter - tracks whether the original middleware function
+   * was dispatched by the router originally
+   * @param {string?} layerPath - if present, provides additional data from the
+   * router about the routed path which the middleware is attached to
+   */
+  private _patchLayer(
+    middlewareLayer: KoaMiddleware,
+    isRouter: boolean,
+    layerPath?: string
+  ): KoaMiddleware {
+    if (middlewareLayer[kLayerPatched] === true) return middlewareLayer;
+    middlewareLayer[kLayerPatched] = true;
+    this._logger.debug('patching Koa middleware layer');
+    return async (context: KoaContext, next: koa.Next) => {
+      if (this._tracer.getCurrentSpan() === undefined) {
+        return await middlewareLayer(context, next);
```

Comment:
I think you don't need the async/await here, since `middlewareLayer` should be async anyway you could just `return middlewareLayer(context, next)` and remove the `async` keyword above.

Reply:
I agree you do not need the await here when you are not adding tracing and just returning `middlewareLayer(context, next)` (line 137). The await on line 148 is necessary though for the span timing to fully capture any asynchronous processes related to the middleware, so I will need to keep the async keyword above.